lib/kaur/result.ex

```diff
@@ -266,30 +314,85 @@ defmodule Kaur.Result do
       iex> Kaur.Result.sequence([Kaur.Result.ok(42), Kaur.Result.error("oops"), Kaur.Result.ok(1337)])
       {:error, "oops"}
   """
-  @spec sequence([result_tuple]) :: ({:ok, [any()]}|{:error, any()})
+  @spec sequence([t(any, any)]) :: t(any, any)
   def sequence(list) do
     case Enum.reduce_while(list, [], &do_sequence/2) do
       {:error, _} = error -> error
-      result -> ok(Enum.reverse result)
+      result -> ok(Enum.reverse(result))
     end
   end
 
+  @doc ~S"""
+  Calls the next function only if it receives an ok tuple but discards the result. It always returns
+  the original tuple.
+
+  ## Examples
+
+      iex> some_logging = fn(x) -> IO.puts "Success #{x}" end
+      ...> {:ok, 42} |> Kaur.Result.tap(some_logging)
+      {:ok, 42}
+
+      iex> some_logging = fn(_) -> IO.puts "Not called logging" end
+      ...> {:error, "oops"} |> Kaur.Result.tap(some_logging)
+      {:error, "oops"}
+  """
+  @spec tap(t(A, B), (A -> any)) :: t(A, B)
+  def tap(data, function), do: map(data, &Kaur.tap(&1, function))
+
+  @doc ~S"""
+  Calls the next function only if it receives an error tuple but discards the result. It always returns
+  the original tuple.
+
+  ## Examples
+
+    iex> some_logging = fn(x) -> IO.puts "Failed #{x}" end
+    ...> {:error, "oops"} |> Kaur.Result.tap_error(some_logging)
+    {:error, "oops"}
+
+    iex> some_logging = fn(_) -> IO.puts "Not called logging" end
+    ...> {:ok, 42} |> Kaur.Result.tap_error(some_logging)
+    {:ok, 42}
+  """
+  @spec tap_error(t(A, B), (B -> any)) :: t(A, B)
+  def tap_error(data, function), do: map_error(data, &Kaur.tap(&1, function))
+
   @doc ~S"""
   Returns the content of an ok tuple if the value is correct. Otherwise it returns the
   default value.
 
   ### Examples
 
-      iex> 42 |> Kaur.Result.ok |> Kaur.Result.with_default(1337)
+      iex> 42 |> Kaur.Result.ok() |> Kaur.Result.with_default(1337)
       42
 
-      iex> "oops" |> Kaur.Result.error |> Kaur.Result.with_default(1337)
+      iex> "oops" |> Kaur.Result.error() |> Kaur.Result.with_default(1337)
       1337
   """
-  @spec with_default(result_tuple, any) :: any
+  @spec with_default(t(any, any), any) :: any
   def with_default({:ok, data}, _default_data), do: data
   def with_default({:error, _}, default_data), do: default_data
 
+  @doc ~S"""
+  Wraps the argument in a `result_tuple`.
+
+  If the argument is already a `result_tuple`, returns the `result_tuple`.
+
+  ### Examples
+
+      iex> 42 |> Kaur.Result.ok() |> Kaur.Result.wrap()
+      {:ok, 42}
+
+      iex> 42 |> Kaur.Result.wrap()
+      {:ok, 42}
+
+      iex> "oops" |> Kaur.Result.error() |> Kaur.Result.wrap()
+      {:error, "oops"}
+  """
+  @spec wrap(any) :: t(any, any)
+  def wrap({:ok, _} = value), do: value
+  def wrap({:error, _} = value), do: value
+  def wrap(value), do: ok(value)
```

Comment:
Shouldn't it call `from_value/1`?

Reply:
Not IMHO :)

Reply:
I'm not for the usage of "from value", because I think that we should avoid the "implicits". 
Imho, it is not the role of wrap to make the arbitrary decision of "nil is rejectable".

Reply:
I agree with @xvw now that he showed me https://hexdocs.pm/elixir/List.html#wrap/1